Samples/src/VisualDebugger/VisualDebugger.cs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.AspNetCore;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Quantum.Simulation.Core;
+using Microsoft.Quantum.Simulation.Simulators;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace vis_sim
+{
+    internal class VisualDebugger
+    {
+        private readonly QuantumSimulator simulator;
+        private readonly StateDumper stateDumper;
+        private readonly IWebHost host;
+        private readonly IHubContext<VisualHub> context;
+        private readonly ManualResetEvent advanceEvent = new ManualResetEvent(true);
+        private readonly IList<(string method, object[] args)> history = new List<(string, object[])>();
+
+        public VisualDebugger(QuantumSimulator simulator)
+        {
+            simulator.OnOperationStart += OnOperationStartHandler;
```

Comment:
You should check that simulator != null and throw `ArgumentNullException` at top.